server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -553,19 +554,21 @@ public void executeQueryPhase(ShardSearchRequest request, SearchShardTask task,
             : "empty responses require more than one shard";
         final IndexShard shard = getShard(request);
         rewriteAndFetchShardRequest(shard, request, listener.delegateFailure((l, orig) -> {
+            final ReaderContext readerContext = request.readerId() == null ? null : findReaderContext(orig.readerId(), orig);
+            IndexService indexService = indicesService.indexServiceSafe(request.shardId().getIndex());
             // check if we can shortcut the query phase entirely.
             if (orig.canReturnNullResponseIfMatchNoDocs()) {
                 assert orig.scroll() == null;
-                final CanMatchShardResponse canMatchResp;
-                try {
-                    ShardSearchRequest clone = new ShardSearchRequest(orig);
-                    canMatchResp = canMatch(clone, false);
+                CanMatchShardResponse canMatchResp;
+                ShardSearchRequest clone = new ShardSearchRequest(orig);
+                try (CanMatchContext canMatchContext = new CanMatchContext(clone, indexService, readerContext, getKeepAlive(request))) {
+                    canMatchResp = canMatch(canMatchContext, false);
                 } catch (Exception exc) {
-                    l.onFailure(exc);
-                    return;
+                    // if can_match throws for some reason, we go ahead with the query phase
+                    canMatchResp = new CanMatchShardResponse(true, null);
```

Comment:
I added a targeted catch exception for sorting related exceptions, but we still need here a broader one. Based on my proposal below to handle all exceptions in the can match method, perhaps this can go away.

Reply:
I removed this, given that exception handling is now in callers code.